evm/p2p/auth.py

```diff
@@ -0,0 +1,284 @@
+import asyncio
+import logging
+import os
+import random
+import struct
+
+import rlp
+from rlp import sedes
+
+from evm.ecc import get_ecc_backend
+from evm.utils.ecdsa import (
+    decode_signature,
+    encode_signature,
+)
+from evm.utils.keccak import (
+    keccak,
+    keccak_updateable,
+)
+from evm.utils.numeric import (
+    int_to_big_endian,
+)
+from evm.p2p import ecies
+from evm.p2p.constants import (
+    auth_ack_len,
+    auth_msg_len,
+    encrypted_auth_ack_len,
+    encrypted_auth_msg_len,
+    hash_len,
+    pubkey_len,
+    signature_len,
+    supported_rlpx_version,
+)
+from evm.p2p.exceptions import AuthenticationError
+from evm.p2p.peer import Peer
+from evm.p2p.utils import (
+    ascii_chr,
+    sxor,
+)
+from evm.utils.secp256k1 import (
+    encode_raw_public_key,
+    private_key_to_public_key,
+)
+
+
+@asyncio.coroutine
+def handshake(remote, privkey):
+    """Perform the auth handshake with the given remote and return a Peer ready to be used.
+
+    The Peer will be configured with the shared secrets established during the handshake.
+    """
+    initiator = HandshakeInitiator(remote, privkey)
+    reader, writer = yield from initiator.connect()
+
+    initiator_nonce = keccak(int_to_big_endian(random.randint(0, 2 ** 256 - 1)))
+    auth_msg = initiator.create_auth_message(initiator_nonce)
+    auth_init = initiator.encrypt_auth_message(auth_msg)
+    writer.write(auth_init)
+
+    ack_data = yield from reader.read(encrypted_auth_ack_len)
+
+    ephemeral_pubkey, responder_nonce, auth_ack = initiator.decode_auth_ack_message(ack_data)
+    aes_secret, mac_secret, egress_mac, ingress_mac = initiator.derive_secrets(
+        initiator_nonce, responder_nonce, ephemeral_pubkey, auth_init, auth_ack)
+
+    peer = Peer(remote, privkey, reader, writer, aes_secret,
```

Comment:
I've always been a bit frightened of function calls and object instantaition that uses more than 3-4 positional arguments given the ease of transposing them or just plain getting the order wrong.  Thoughts on converting this to use keyword arguments?

Reply:
Yeah, I'm also a bit weary of them so I'm happy to do that. But I'll move this whole function and the creation of the Peer class to another branch I have, as this really belongs there and will mean less conflicts when I rebase the other branch after we're done with the code review